linux/mcompile/keymap/mc_import_rules.cpp

```diff
@@ -0,0 +1,651 @@
+  Related Files:    
+  Dependencies:     
+
+  Bugs:             
+  Todo:             
+  Notes:            
+  History:          03 Aug 2014 - mcdurdin - I4327 - V9.0 - Mnemonic layout compiler follow-up
+                    03 Aug 2014 - mcdurdin - I4353 - V9.0 - mnemonic layout recompiler mixes up deadkey rules
+                    31 Dec 2014 - mcdurdin - I4550 - V9.0 - logical flaw in mnemonic layout recompiler means that AltGr base keys are never processed
+                    06 Feb 2015 - mcdurdin - I4552 - V9.0 - Add mnemonic recompile option to ignore deadkeys
+*/
+
+#include <vector>
+#include <string>
+#include <stdio.h>
+//#include "km_types.h"
+#include "mc_kmxfile.h"
+#include "keymap.h"
+
+const int KMX_ShiftStateMap[] = {
+  ISVIRTUALKEY,
+  ISVIRTUALKEY | K_SHIFTFLAG,
+  ISVIRTUALKEY | K_CTRLFLAG,
+  ISVIRTUALKEY | K_SHIFTFLAG | K_CTRLFLAG,
+  0,
+  0,
+  ISVIRTUALKEY | RALTFLAG,
+  ISVIRTUALKEY | RALTFLAG | K_SHIFTFLAG,
+  0,
+  0
+};
+
+DeadKey::DeadKey(KMX_WCHAR deadCharacter) {
+  this->m_deadchar = deadCharacter;
+}
+
+KMX_WCHAR DeadKey::KMX_DeadCharacter() {
+  return this->m_deadchar;
+}
+
+void DeadKey::KMX_AddDeadKeyRow(KMX_WCHAR baseCharacter, KMX_WCHAR combinedCharacter) {
+  this->m_rgbasechar.push_back(baseCharacter);
+  this->m_rgcombchar.push_back(combinedCharacter);
+}
+
+bool DeadKey::KMX_ContainsBaseCharacter(KMX_WCHAR baseCharacter) {
+  std::vector<KMX_WCHAR>::iterator it;
+  for(it=this->m_rgbasechar.begin(); it<m_rgbasechar.end(); it++) {
+    if(*it == baseCharacter) {
+      return true;
+    }
+  }
+  return false;
+}
+
+int KMX_ToUnicodeEx(guint keycode, PKMX_WCHAR pwszBuff, int shift_state_pos, int caps,GdkKeymap *keymap) {
+  GdkKeymapKey *maps;
+  guint *keyvals;
+  gint count;
+
+  if (!gdk_keymap_get_entries_for_keycode(keymap, keycode, &maps, &keyvals, &count))
+    return 0;
+
+  if (!(shift_state_pos <= count))
+   return 0;
+
+  if (!(keycode <= keycode_max))
+    return 0;
+
+  KMX_DWORD KeyVal= (KMX_DWORD) KMX_get_KeyVal_From_KeyCode(keymap, keycode, ShiftState(shift_state_pos), caps);
+  std::u16string str = convert_DeadkeyValues_To_U16str(KeyVal);
+  pwszBuff[0]= * (PKMX_WCHAR)  str.c_str();
+
+
+  g_free(keyvals);
+  g_free(maps);
+
+  if((KeyVal >=  deadkey_min) && (KeyVal <=  deadkey_max))          // deadkeys
+    return -1;
+  else if(gdk_keyval_to_unicode(KeyVal)  == 0)                      // NO UNICODE
+    return 0;
+  else                                                              // usable char
+    return 1;
+}
+
+int KMX_DeadKeyMap(int index, std::vector<DeadKey *> *deadkeys, int deadkeyBase, std::vector<KMX_DeadkeyMapping> *deadkeyMappings) {   // I4327   // I4353
+  for(size_t i = 0; i < deadkeyMappings->size(); i++) {
+    if((*deadkeyMappings)[i].deadkey == index) {
+      return (*deadkeyMappings)[i].dkid;
+    }
+  }
+
+  for(size_t i = 0; i < deadkeys->size(); i++) {
+    if((*deadkeys)[i]->KMX_DeadCharacter() == index) {
+      return deadkeyBase + i;
```

Comment:
Sounds like a question for @mcdurdin then :smile:

Reply:
This is because there may be existing deadkeys in the keyboard, so we need to make sure we add any new deadkeys with unique indices after all the existing deadkeys.